### PR TITLE
Segregated List MMU 

### PR DESCRIPTION
**Pull request by qingsenzhang**

### CODE REVIEW

File: include/mxnet/mem_mgr.h [hunk 1]
```diff
@@ -3,23 +3,31 @@
...
 #include <mutex>
+#include <math.h>
 #include <stdio.h>
 #include <string>
 
 namespace mxnet {
 
 const int MINALLOCSIZE_ = 128;
-const int FREELISTSIZE_ = 21;
+const double GPUUTILRATIO = 0.95; //save some memory for each device(subject to change)
```
Comment (fegin) [inline include/mxnet/mem_mgr.h]: MINALLOCSIZE_-> MIN_ALLOC_SIZE
GPUUTILRATIO -> GPU_UTIL_RATIO

File: include/mxnet/mem_mgr.h [hunk 2]
```diff
@@ -30,70 +38,104 @@ inline std::string memGetStatusString(memStatus_t status) {
...
-  //prevent seg fault if index is out of range
-  idx = (idx < 0) ? 0 : idx;
-  idx = (idx > 20) ? 20 : idx;
-  return idx;
+inline int nextPowerOfTwo(int size) {
+  int result = pow(2, ceil(log(size) / log(2)));
+  return result;
+}
+
+inline int getListIdx(size_t size) {
```
Comment (fegin) [inline include/mxnet/mem_mgr.h]: For the camel case naming, the first character of function names should be uppercase (Google coding style). Please change all functions.

File: include/mxnet/mem_mgr.h [hunk 3]
```diff
@@ -30,70 +38,104 @@ inline std::string memGetStatusString(memStatus_t status) {
...
+  private:
+    Block* start_;
+    Block** freeList_;
+    size_t total_;
+    size_t allocated_;
+    size_t free_;
+    int freeListSize_;
+    int gpuIdx_;
+   
+  private:
```
Comment (fegin) [inline include/mxnet/mem_mgr.h]: Redundant private keyword.

File: include/mxnet/mem_mgr.h [hunk 4]
```diff
@@ -30,70 +38,104 @@ inline std::string memGetStatusString(memStatus_t status) {
...
+    
+  public:
+    size_t getTotal() { return total_; }
+    size_t getFree() { return free_; }
+    size_t getAllocated() { return allocated_; }  
+    int getFreeListSize() { return freeListSize_; }
+    Block** getFreeList() { return freeList_; }
+    MemoryPool getMemPool() { return memPool_; }  
+ 
+  public:
```
Comment (fegin) [inline include/mxnet/mem_mgr.h]: Redundant public keyword.

File: include/mxnet/mem_mgr.h [hunk 5]
```diff
@@ -30,70 +38,104 @@ inline std::string memGetStatusString(memStatus_t status) {
...
+    void* Alloc(size_t size);
+    cudaError_t Free(void* ptr); 
+    
+  private:
```
Comment (fegin) [inline include/mxnet/mem_mgr.h]: Please group all private members together.

File: include/mxnet/mem_mgr.h [hunk 6]
```diff
@@ -30,70 +38,104 @@ inline std::string memGetStatusString(memStatus_t status) {
...
     static std::shared_ptr<MemoryManager> _GetSharedRef();
     cudaError_t Malloc(void*& devptr, size_t size, int deviceIdx);
     cudaError_t Free(void* devptr, int deviceIdx);
-    cudaError_t Memcpy(int deviceId, void* dst, 
+    cudaError_t Memcpy(int deviceIdx, void* dst, 
                        const void* src, size_t count, enum cudaMemcpyKind kind);
     cudaError_t MemGetInfo(int deviceId, size_t* total, size_t* free);   
     bool TryAllocate(int deviceId, size_t size);
 
   private:
```
Comment (fegin) [inline include/mxnet/mem_mgr.h]: Please group all private members together.

File: src/storage/mem_mgr.cc [hunk 7]
```diff
@@ -28,125 +152,87 @@ std::shared_ptr<MemoryManager> MemoryManager::_GetSharedRef() {
...
+    
+    size_t avail, total;
+    size_t mb = 1 << 20;
+    CUDA_CALL(cudaMemGetInfo(&avail, &total));
+  
+    avail = static_cast<size_t>(avail * GPUUTILRATIO);  
+    char* wholeMemory = NULL;
+    while (cudaMalloc((void**)&wholeMemory, avail) == cudaErrorMemoryAllocation) {
+        avail -= mb;
+        if (avail <= 0) break;
```
Comment (fegin) [inline src/storage/mem_mgr.cc]: There should be a warning or assert message here.

File: src/storage/mem_mgr.cc [hunk 8]
```diff
@@ -28,125 +152,87 @@ std::shared_ptr<MemoryManager> MemoryManager::_GetSharedRef() {
...
-//}
+MemoryManager::~MemoryManager() {
+  typedef std::map<char*, Block*> MemoryPool;
+  for (int deviceIdx = 0; deviceIdx < deviceCount_; deviceIdx++) {
+    CUDA_CALL(cudaSetDevice(deviceIdx));
+    BuddySystem* buddy = buddy_[deviceIdx];
+    MemoryPool mp = buddy->getMemPool();
+    while (!mp.empty()) {
+      buddy->Free((void*)(mp.begin()->first));
+    }
```
Comment (fegin) [inline src/storage/mem_mgr.cc]: For completeness, you should also call cudaFree after this while.

File: src/storage/mem_mgr.cc [hunk 9]
```diff
@@ -28,125 +152,87 @@ std::shared_ptr<MemoryManager> MemoryManager::_GetSharedRef() {
...
+    MemoryPool mp = buddy->getMemPool();
+    while (!mp.empty()) {
+      buddy->Free((void*)(mp.begin()->first));
+    }    
+  }
+}
 
 cudaError_t MemoryManager::Malloc(void*& devptr, size_t size, int deviceIdx) {
   std::lock_guard<std::mutex> lock(mutex_);
   CUDA_CALL(cudaSetDevice(deviceIdx));
```
Comment (fegin) [inline src/storage/mem_mgr.cc]: Is this API call (cudaSetDevice) still needed?

File: src/storage/mem_mgr.cc [hunk 10]
```diff
@@ -28,125 +152,87 @@ std::shared_ptr<MemoryManager> MemoryManager::_GetSharedRef() {
...
-  curr->setNext(freeList_[idx]);
-  freeList_[idx] = curr;
+  buddy_[deviceIdx]->Free(devptr);
   return cudaSuccess;
 }
         
-cudaError_t MemoryManager::Memcpy(int deviceId, void* dst, const void* src, size_t count, enum cudaMemcpyKind kind) {
+cudaError_t MemoryManager::Memcpy(int deviceIdx, void* dst, const void* src, size_t count, enum cudaMemcpyKind kind) {
   //TODO(qingsen): need implementation
+  std::lock_guard<std::mutex> lock(mutex_);
```
Comment (fegin) [inline src/storage/mem_mgr.cc]: Why do we need a mutex here? I believe cudaMemcpy is thread-safe.

**Review by fegin — CHANGES_REQUESTED** [review]

**Comment by Sotskin** [comment]
Looks good.

**Review by Sotskin — APPROVED** [review]

**Comment by fegin** [comment]
Why do you merge this to gpu_swap_mem since you merge gpu_swap? gpu_swap_mem is apparently too old. So there are too many changes after you merged gpu_swap. Please close this pull request and create another pull request which will be directly merged to gpu_swap.

**Pull request closed without merging**